### PR TITLE
Enable pytorch nightly CI

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

- Make necessary changes to build `huggingface/transformers-pytorch-nightly-gpu`
- Update `self-nightly-scheduled.yml` to run PyTorch nightly build CI (almost a copy from scheduled CI)

[test workflow run](https://github.com/huggingface/transformers/actions/runs/2354046716) here.
[docker build run](https://github.com/huggingface/transformers/actions/runs/2353866659)

#### Print versions
Screenshot 2022-05-19 214304

### CODE REVIEW

File: docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+RUN python3 -m pip install --no-cache-dir -e ./transformers[deepspeed-testing]
+
+# Install nightly build PyTorch
+RUN python3 -m pip uninstall -y torch torchvision torchaudio
+RUN python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu102
+
+# Currently not working with nightly-built PyTorch.
+# (In fact, this step doesn't work if we do `python3 -m pip install --no-cache-dir -U torch torchvision torchaudio`)
+# RUN git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed && rm -rf build && \
+#     DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install -e . --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check 2>&1
```
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile:21]: To investigate

File: .github/workflows/self-nightly-scheduled.yml [hunk 2]
```diff
@@ -18,233 +18,208 @@ env:
...
+        working-directory: /workspace/transformers
+        run: git fetch && git checkout ${{ github.sha }}
+
+#      (not working with the docker image `transformers-pytorch-deepspeed-nightly-gpu`)
+#      - name: Re-compile DeepSpeed
+#        run: |
+#          pip install deepspeed # installs the deps correctly
+#          rm -rf DeepSpeed
+#          git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed && rm -rf build
+#          DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install -e . --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check
```
Comment (ydshieh) [inline .github/workflows/self-nightly-scheduled.yml:182]: To investigate. But without this, we get a lot of
```
'DeepSpeedCPUAdam' object has no attribute 'ds_opt_adam'
```

**Comment by ydshieh** [comment]
@stas00, @LysandreJik 

I run the docker image `transformers-pytorch-deepspeed-latest-gpu` and found it's `PyTroch` has version `1.9.0`.
This image is based on `nvcr.io/nvidia/pytorch:21.03-py3`, which is used in the job `Test Torch CUDA Extensions` for both daily scheduled CI and push CI.

We can discuss about this. But so far I won't include DeepSpeed test with nightly-built PyTorch.

```
>>> import torch
>>> torch.__version__
'1.9.0a0+df837d0'
>>> exit()
```

File: docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,27 @@
+FROM nvcr.io/nvidia/pytorch:21.03-py3
```
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile:1]: not used

**Comment by stas00** [comment]
> I run the docker image transformers-pytorch-deepspeed-latest-gpu and found it's PyTroch has version 1.9.0.
This image is based on nvcr.io/nvidia/pytorch:21.03-py3, which is used in the job Test Torch CUDA Extensions for both daily scheduled CI and push CI.

But that's not nightly. I don't think you can rely on any pre-made docker to run nightly. It has to be manually installed since it is a new version every day. You probably want to switch to [22.04](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_22-04.html#rel_22-04) (latest at the moment) and then update it to the actual nightly. Does it make sense?

I also don't understand why deepspeed tests were removed. It's critical that we run deepspeed tests on nightly.

**Comment by stas00** [comment]
> Should we use the latest stable Pytorch instead?

Yes, please.

I think we should use the latest stable pytorch for all our tests unless we explicitly test older pytorch versions.

> I even tried with PyTorch Stable + DeepSpeed and the docker image also fails.

Could you please point me to the actual issues and I will help to sort them out? 

We can discuss it on slack if it's easier.

File: .github/workflows/self-nightly-scheduled.yml [hunk 4]
```diff
@@ -18,233 +18,211 @@ env:
...
+      - name: Update clone
+        working-directory: /workspace/transformers
+        run: git fetch && git checkout ${{ github.sha }}
+
+      # (not working with the docker image `transformers-pytorch-deepspeed-nightly-gpu`)
+      - name: Re-compile DeepSpeed
+        run: |
+          pip install deepspeed # installs the deps correctly
+          rm -rf DeepSpeed
+          git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed && rm -rf build
```
Comment (stas00) [inline .github/workflows/self-nightly-scheduled.yml:181]:
```suggestion
          git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed 
```

the `build` won't be there if it's fresh clone, so it's not needed.

**Review by stas00 — APPROVED** [review]
Looks good, thank you, @ydshieh!

File: .github/workflows/self-nightly-scheduled.yml (hunk 4 above)
Comment (ydshieh) [reply .github/workflows/self-nightly-scheduled.yml:181]: actually we build this in the docker image, but also in the test job

https://github.com/huggingface/transformers/blob/52b5f515a4c59b6c460b2c440a35451d807ecf26/docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile#L18-L19

I guess it's probably redundant, and we only need to keep one. Would let @LysandreJik to judge.

File: docker/transformers-pytorch-gpu/Dockerfile [hunk 5]
```diff
@@ -14,7 +14,9 @@ RUN python3 -m pip install --no-cache-dir -e ./transformers[dev-torch,testing]
 # If set to nothing, will install the latest version
 ARG PYTORCH=''
 
-RUN [ ${#PYTORCH} -gt 0 ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'; python3 -m pip install --no-cache-dir -U $VERSION
+RUN [ ${#PYTORCH} -gt 0 -a ${#PYTORCH} != "pre" ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'
+RUN [ ${#PYTORCH} != "pre" ] && python3 -m pip install --no-cache-dir -U $VERSION || python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu102
+
```
Comment (ydshieh) [inline docker/transformers-pytorch-gpu/Dockerfile:19]: Need to be fixed, sorry

File: docker/transformers-pytorch-gpu/Dockerfile [hunk 6]
```diff
@@ -12,9 +12,17 @@ RUN git clone https://github.com/huggingface/transformers && cd transformers &&
...
 
 # If set to nothing, will install the latest version
+# `pre` means `nightly built PyTorch`
 ARG PYTORCH=''
 
-RUN [ ${#PYTORCH} -gt 0 ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'; python3 -m pip install --no-cache-dir -U $VERSION
+RUN [ ${#PYTORCH} -gt 0 -a ${#PYTORCH} != "pre" ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'
+RUN [ ${#PYTORCH} != "pre" ] && python3 -m pip install --no-cache-dir -U $VERSION || python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu102
+
+# Install extra modules for nightly PyTorch CI
```
Comment (stas00) [inline docker/transformers-pytorch-gpu/Dockerfile:21]:
```suggestion
# Install torchdynamo dependencies
```

Let's be more specific of why these are installed.

**Review by LysandreJik — APPROVED** [review]
Looks good to me, @ydshieh! Thanks for taking care of that.

**Review by ydshieh — COMMENTED** [review]
Hi @LysandreJik @stas00 

After the recent changes regarding tests, I have to update this PR. In particular, **I would like to make sure the goal of the scheduled nightly test(s) are**:
  - tests with nightly torch + latest TF (+ stable Flax)
  - tests with latest torch + nightly TF (+ stable Flax)
  - tests with latest torch + latest TF (+ nightlyFlax)

Here the summary of the latest changes in this PR
- build a new docker image with **nightly torch + latest stable TF**
- make `self-nightly-scheduled` the same as `self-scheduled`, except:
  - only keep the following tests (so far), running with docker images having **nightly torch and/or deepspeed + stable release TF**
    - run_tests_single_gpu
    - run_tests_multi_gpu
    - run_all_tests_torch_cuda_extensions_gpu
  - (we could add **`run_examples_gpu` and `run_pipelines_torch_gpu`** in a follow up PR if necessary)

File: docker/transformers-all-latest-gpu/Dockerfile [hunk 7]
```diff
@@ -3,6 +3,9 @@ LABEL maintainer="Hugging Face"
 
 ARG DEBIAN_FRONTEND=noninteractive
 
+# Used to read variables from `~/.profile` (to pass dynamic created variables between RUN commands)
+SHELL ["sh", "-lc"]
```
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:7]: Dockfile is really limited, and I really need to pass some variables between `RUN` command.

File: docker/transformers-all-latest-gpu/Dockerfile [hunk 8]
```diff
@@ -21,11 +24,20 @@ ARG REF=main
 RUN git clone https://github.com/huggingface/transformers && cd transformers && git checkout $REF
 RUN python3 -m pip install --no-cache-dir -e ./transformers[dev,onnxruntime]
 
-RUN python3 -m pip install --no-cache-dir -U torch==$PYTORCH torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/$CUDA
+# TODO: Handle these in a python utility script
+RUN [ ${#PYTORCH} -gt 0 -a "$PYTORCH" != "pre" ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'; echo "export VERSION='$VERSION'" >> ~/.profile
```
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:28]: This is where we need `SHELL ["sh", "-lc"]` in order to pass `VERSION` to the next line.

File: docker/transformers-all-latest-gpu/Dockerfile [hunk 9]
```diff
@@ -21,11 +24,20 @@ ARG REF=main
...
+RUN echo torch=$VERSION
+# `torchvision` and `torchaudio` should be installed along with `torch`, especially for nightly build.
+# Currently, let's just use their latest releases (when `torch` is installed with a release version)
+# TODO: We might need to specify proper versions that work with a specific torch version (especially for past CI).
+RUN [ "$PYTORCH" != "pre" ] && python3 -m pip install --no-cache-dir -U $VERSION torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/$CUDA || python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/$CUDA
```
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:33]: I really think we can make things and life easier here. I have some POC, and can do it in a follow-up PR.

File: docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+RUN python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/$CUDA
+
+RUN python3 -m pip install --no-cache-dir ./transformers[deepspeed-testing]
+
+# Pre-build **nightly** release of DeepSpeed, so it would be ready for testing (otherwise, the 1st deepspeed test will timeout)
+RUN python3 -m pip uninstall -y deepspeed
+# This has to be run inside the GPU VMs running the tests. (So far, it fails here due to GPU checks during compilation.)
+# Issue: https://github.com/microsoft/DeepSpeed/issues/2010
+# RUN git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed && rm -rf build && \
+#    DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install . --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check 2>&1
```
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile:28]: See https://github.com/microsoft/DeepSpeed/issues/2010

File: docker/transformers-pytorch-deepspeed-latest-gpu/Dockerfile [hunk 11]
```diff
@@ -13,13 +16,16 @@ RUN git clone https://github.com/huggingface/transformers && cd transformers &&
...
 RUN python3 -m pip install --no-cache-dir ./transformers[deepspeed-testing]
 
-# Pre-build DeepSpeed, so it would be ready for testing (otherwise, the 1st deepspeed test will timeout)
+# Pre-build **latest** DeepSpeed, so it would be ready for testing (otherwise, the 1st deepspeed test will timeout)
 RUN python3 -m pip uninstall -y deepspeed
-RUN DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install deepspeed --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check
+# This has to be run (again) inside the GPU VMs running the tests.
+# The installation works here, but some tests fail, if we don't pre-build deepspeed again in the VMs running the tests.
+# TODO: Find out why test fail.
+RUN DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install deepspeed --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check 2>&1
```
Comment (ydshieh) [inline docker/transformers-pytorch-deepspeed-latest-gpu/Dockerfile:28]: need to pre-build again in workflow file (i.e. when docker is launched) - until a fix in deepspeed is done.

**Review by ydshieh — COMMENTED** [review]
Hey, @LysandreJik @stas00 !

I have updated this PR. I think the best way to save your time reviewing is to check my comments below.

If everything is fine, I will run some tests to make sure before merge.

File: .github/workflows/self-nightly-scheduled.yml [hunk 12]
```diff
@@ -1,250 +1,235 @@
-name: Self-hosted runner; Nightly (scheduled)
+name: Self-hosted runner (nightly)
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply .github/workflows/self-nightly-scheduled.yml:1]: well, I can change this back - I didn't pay attention here.

File: docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile (hunk 10 above)
Comment (stas00) [reply docker/transformers-pytorch-deepspeed-nightly-gpu/Dockerfile:28]: We should be able to fix this one quickly I think on the DS side - let's see what the maintainers think - I can make a PR to not have it fail easily.

**Comment by stas00** [comment]
Looks good to me, @ydshieh! Thank you!

File: .github/workflows/self-nightly-scheduled.yml [hunk 13]
```diff
@@ -1,250 +1,236 @@
...
+        run: git fetch && git checkout ${{ github.sha }}
+
+      # To avoid unknown test failures
+      - name: Pre build DeepSpeed *again*
+        working-directory: /workspace
+        run: |
+          python3 -m pip uninstall -y deepspeed
+          rm -rf DeepSpeed
+          git clone https://github.com/microsoft/DeepSpeed && cd DeepSpeed && rm -rf build
+          DS_BUILD_CPU_ADAM=1 DS_BUILD_AIO=1 DS_BUILD_UTILS=1 python3 -m pip install . --global-option="build_ext" --global-option="-j8" --no-cache -v --disable-pip-version-check
```
Comment (ydshieh) [inline .github/workflows/self-nightly-scheduled.yml:188]: pre-build deepspeed (master version), which is currently necessary.

**Comment by ydshieh** [comment]
A full workflow run is [here](https://github.com/huggingface/transformers/actions/runs/2490446459).

**Pull request merged**